R/bmerge.R

```diff
@@ -84,7 +84,19 @@ bmerge = function(i, x, icols, xcols, roll, rollends, nomatch, mult, ops, verbos
           next
         }
       }
-      stopf("Incompatible join types: %s (%s) and %s (%s). Factor columns must join to factor or character columns.", xname, x_merge_type, iname, i_merge_type)
+      condition_message = gettextf("Incompatible join types: %s (%s) and %s (%s). Factor columns must join to factor or character columns.", xname, x_merge_type, iname, i_merge_type)
+      condition = structure(
+        list(
+          message = condition_message,
+          call = sys.call(sys.nframe()-1L),
+          bmerge_x_arg_col_name      = names(x)[xcol],
+          bmerge_x_arg_type          = x_merge_type,
+          bmerge_i_arg_col_name      = names(i)[icol],
+          bmerge_i_arg_type          = i_merge_type
+        ),
+        class = c("dt_bmerge_incompatible_type_error", "error", "condition")
+      )
```

Comment:
`structure()` is weirdly slow; better to just add `class()` as a follow-up step

```suggestion
      condition = list(
        message = condition_message,
        call = sys.call(sys.nframe()-1L),
        bmerge_x_arg_col_name      = names(x)[xcol],
        bmerge_x_arg_type          = x_merge_type,
        bmerge_i_arg_col_name      = names(i)[icol],
        bmerge_i_arg_type          = i_merge_type
      )
      class(condition) = c("dt_bmerge_incompatible_type_error", "error", "condition")
```